### PR TITLE
add form_data with query

### PR DESCRIPTION
**Pull request by eschutho**
### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/explore/form_data/commands/create.py [hunk 1]
```diff
@@ -39,20 +39,23 @@ def __init__(self, cmd_params: CommandParameters):
...
+            datasource_type = self._cmd_params.datasource_type
             chart_id = self._cmd_params.chart_id
             tab_id = self._cmd_params.tab_id
             actor = self._cmd_params.actor
             form_data = self._cmd_params.form_data
-            check_access(dataset_id, chart_id, actor)
-            contextual_key = cache_key(session.get("_id"), tab_id, dataset_id, chart_id)
+            check_access(datasource_id, chart_id, actor, datasource_type)
+            contextual_key = cache_key(
+                session.get("_id"), tab_id, datasource_id, chart_id
```
Comment (eschutho) [inline superset/explore/form_data/commands/create.py:50]: missing datasource_type here

File: superset/explore/form_data/commands/parameters.py [hunk 2]
```diff
@@ -23,7 +23,8 @@
 @dataclass
 class CommandParameters:
     actor: User
-    dataset_id: int = 0
+    datasource_type: str
```
Comment (eschutho) [inline superset/explore/form_data/commands/parameters.py:26]: use the DatasourceType enum

File: tests/integration_tests/explore/form_data/api_tests.py [hunk 3]
```diff
@@ -89,6 +99,30 @@ def test_post(client, chart_id: int, dataset_id: int):
...
+        "datasource": f"{query_id}__query",
+    }
+    payload = {
+        "datasource_id": query_id,
+        "chart_id": chart_id,
+        "form_data": json.dumps(form_data),
+        "datasource_type": "query",
+    }
+    resp = client.post("api/v1/explore/form_data", json=payload)
+    print(resp.data)
```
Comment (hughhhh) [inline tests/integration_tests/explore/form_data/api_tests.py:122]: remove this

File: superset/charts/schemas.py [hunk 4]
```diff
@@ -199,7 +199,7 @@ class ChartPostSchema(Schema):
     datasource_id = fields.Integer(description=datasource_id_description, required=True)
     datasource_type = fields.String(
         description=datasource_type_description,
-        validate=validate.OneOf(choices=("druid", "table", "view")),
+        validate=validate.OneOf(choices=("druid", "table", "view", "query")),
```
Comment (hughhhh) [inline superset/charts/schemas.py:202]: can we change these to be the enums as well?

File: scripts/tests/run.sh [hunk 5]
```diff
@@ -26,19 +26,20 @@ function reset_db() {
   echo --------------------
   echo Reseting test DB
   echo --------------------
-  docker-compose stop superset-tests-worker superset || true
-  RESET_DB_CMD="psql \"postgresql://${DB_USER}:${DB_PASSWORD}@127.0.0.1:5432\" <<-EOF
+  # docker-compose stop superset-tests-worker superset || true
+  psql -U ${DB_USER}
```
Comment (hughhhh) [inline scripts/tests/run.sh:30]: do we want to check in these changes?

**Pull request closed without merging**